### ISSUE TITLE
Limiter (and LimPID) does not treat limits literally

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#987]
**Reported by wischhusen on 14 Feb 2013 15:25 UTC**
Consider the following example:

```
model Limit

  Modelica.Blocks.Continuous.LimPID PID(
    controllerType=Modelica.Blocks.Types.SimpleController.PI,
    Ti=1,
    yMax=1,
    yMin=0,
    initType=Modelica.Blocks.Types.InitPID.InitialOutput,
    xi_start=0,
    y_start=1,
    k=1) annotation (Placement(transformation(extent={{-26,-6},{-6,14}})));
  Modelica.Blocks.Sources.Constant const(k=1)
    annotation (Placement(transformation(extent={{-66,10},{-46,32}})));
  Modelica.Blocks.Sources.Constant const1(k=10)
    annotation (Placement(transformation(extent={{-44,-36},{-24,-16}})));
  Modelica.Blocks.Math.Sqrt sqrt1
    annotation (Placement(transformation(extent={{14,-6},{34,14}})));
equation 
  connect(const.y, PID.u_s) annotation (Line(
      points={{-45,21},{-36.5,21},{-36.5,4},{-28,4}},
      color={0,0,127},
      smooth=Smooth.None));
  connect(const1.y, PID.u_m) annotation (Line(
      points={{-23,-26},{-16,-26},{-16,-8}},
      color={0,0,127},
      smooth=Smooth.None));
  connect(sqrt1.u, PID.y) annotation (Line(
      points={{12,4},{-5,4}},
      color={0,0,127},
      smooth=Smooth.None));
end Limit;

```

The user wanted to limit the output to a minimum value of 0 but not negative (because it was clear that this would lead to an error in a connected model). 

In my Modelica-tool I receive an error as follows:  

The following error was detected at time: 0.112

Model error - sqrt (PID.y) = sqrt (-0.008)

This is caused by the PID-internal limiter block which is having the follwing line:

```
y = smooth(0,if u > uMax then uMax else if u < uMin then uMin else u);
```

Would it be suitable to go with this line instead in order to take limits literally in any case? Or is it a question of efficiency? Or even a tool-problem, because **smooth()** should do the job?

```
y = smooth(0,noEvent(if u > uMax then uMax else if u < uMin then uMin else u));
```

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/987

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 14 Feb 2013 16:04 UTC**
The behavior is correct for smooth(), because smooth may detect events and thus needs to evaluate at the other side of the condition. Only noEvent will give the behavior you want. smooth() can be more efficient. I think you are responsible yourself to guard against evaluation of negative numbers inside your sqrt() call. 

Maybe we would like to discuss a bit more, but I think this ticket is invalid. The behavior is in accordance with the spec, as far as I can see.

**Comment by modelica-trac-importer** [comment]
**Comment by wischhusen on 15 Feb 2013 08:20 UTC**
Replying to [comment:1 hubertus]:

If smooth() is in accordance to the spec than I would suggest to change the limiter (in the way I suggested?) because it is better modelling style with regard to users expectations. The limiter should do what the component is supposed to do. Moreover, errors occur (e.g. divergence) which might not be easy to locate. 

Is there a strong need for using smooth() here? If the user wants to trigger events it could be selected by a Boolean parameter.

I just thought it would be a helpful enhancement (not a defect).

> The behavior is correct for smooth(), because smooth may detect events and thus needs to evaluate at the other side of the condition. Only noEvent will give the behavior you want. smooth() can be more efficient. I think you are responsible yourself to guard against evaluation of negative numbers inside your sqrt() call. 
> 
> Maybe we would like to discuss a bit more, but I think this ticket is invalid. The behavior is in accordance with the spec, as far as I can see.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 30 Mar 2013 16:19 UTC**
Fixed in 92cbbbeadbf3917af3cc0c4f4a51fdcf384a3717:
Included a parameter "strict" in the "Advanced" menu of the following models:
  Modelica.Blocks.Continuous.LimIntegrator
  Modelica.Blocks.Continuous.LimPID
  Modelica.Blocks.Nonlinear.Limiter
  Modelica.Blocks.Nonlinear.VariableLimiter
  Modelica.Blocks.Nonlinear.SlewRateLimiter
The default is "strict=false" (so previous behavior).
If strict=true, all limits are defined with noEvent(..).
